lib/util/basen.h

```diff
@@ -54,165 +54,164 @@ void decode_b64(Iter1 start, Iter1 end, Iter2 out);
     }
 
     static char decode(char c)
     {
         if (c >= '0' && c <= '9') {
             return c - '0';
-        }
-        else if (c >= 'A' && c <= 'F') {
+        } else if (c >= 'A' && c <= 'F') {
             return c - 'A' + 10;
         }
-        return -1;
+        return Error;
     }
 };
 
 struct b32_conversion_traits
 {
-    static int group_length()
+    static size_t group_length()
     {
-        return 5;
+       return 5;
     }
 
     static char encode(unsigned int index)
     {
-        static const char dictionary[33] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
-        assert(index < 32);
+        const char * dictionary = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
+        assert(index < strlen(dictionary));
         return dictionary[index];
     }
 
     static char decode(char c)
     {
         if (c >= 'A' && c <= 'Z') {
             return c - 'A';
-        }
-        else if (c >= '2' && c <= '7') {
+        } else if (c >= '2' && c <= '7') {
             return c - '2' + 26;
         }
-        return -1;
+        return Error;
     }
 };
 
 struct b64_conversion_traits
 {
-    static int group_length()
+    static size_t group_length()
     {
-        return 6;
+       return 6;
     }
 
     static char encode(unsigned int index)
     {
-        static const char dictionary[65] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
-        assert(index < 64);
+        const char* const dictionary = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+        assert(index < strlen(dictionary));
         return dictionary[index];
     }
 
     static char decode(char c)
     {
-        const int indexOf_a = 26;
-        const int indexOf_0 = indexOf_a+ indexOf_a;
-        const int indexOf_Plus = indexOf_0 + 10;
+        const int alph_len = 26;
         if (c >= 'A' && c <= 'Z') {
             return c - 'A';
+        } else if (c >= 'a' && c <= 'z') {
+            return c - 'a' + alph_len * 1;
+        } else if (c >= '0' && c <= '9') {
+            return c - '0' + alph_len * 2;
+        } else if (c == '+') {
+            return c - '+' + alph_len * 2 + 10;
+        } else if (c == '/') {
+            return c - '/' + alph_len * 2 + 11;
         }
-        else if (c >= 'a' && c <= 'z') {
-            return c - 'a' + indexOf_a;
-        }
-        else if (c >= '0' && c <= '9') {
-            return c - '0' + indexOf_0;
-        }
-        else if (c == '+') {
-            return c - '+' + indexOf_Plus;
-        }
-        else if (c == '/') {
-            return c - '/' + indexOf_Plus + 1;
-        }
-        return -1;
+        return Error;
     }
 };
 
 template<class ConversionTraits, class Iter1, class Iter2>
 void decode(Iter1 start, Iter1 end, Iter2 out)
 {
     Iter1 iter = start;
-    int output_current_bit = 0;
+    size_t output_current_bit = 0;
     char buffer = 0;
-    int trailing_equal = 0;
 
     while (iter != end) {
-        if (*iter == '=') {
```

Comment:
I don't understand why this passes tests: if we do not deal with the trailing '=' explicitly, when are we writing to the `out` buffer? It seems we would just trigger the error condition that just skips the rest of the loop

Reply:
It just ignores it in the if (value == Error) part, and then continues to the next iteration.

Previous version does almost the same with the difference than trailing_equal is computed and then used on line 200.

That line does not exist in upstream (and never existed), so it must have been added by you in 860ef61989b955a0f147e797d8895635a2fc65b5. If so, then we need tests to cover case where it is executed and reintroduce these changes here (and make PR upstream).